sklearn/cross_decomposition/_pls.py

```diff
@@ -246,7 +246,7 @@ def fit(self, X, Y):
         # With PLSRegression n_components is bounded by the rank of (X.T X) see
         # Wegelin page 25. With CCA and PLSCanonical, n_components is bounded
         # by the rank of X and the rank of Y: see Wegelin page 12
-        rank_upper_bound = p if self.deflation_mode == "regression" else min(n, p, q)
+        rank_upper_bound = min(n - 1, p) if self.deflation_mode == "regression" else min(n, p, q)
```

Comment:
Oh, I did not see that you make this change here. I see two issues with the rest of the code:

- We break the backward compatibility since we modify the behaviour. If it was not a bug, then this is not worth doing it.
- From what I can read from the comment, `n_components` is bounded by the upper bound of the rank of the matrix `X.T @ X` which is a `(n_features, n_features)` matrix. To me, it makes sense that it does not depend at all on `n_samples` then.